lib/mix/tasks/docs.ex

```diff
@@ -141,8 +140,31 @@ defmodule Mix.Tasks.Docs do
     end
   end
 
-  defp normalize_source_beam(options) do
-    Keyword.put_new(options, :source_beam, Mix.Project.compile_path)
+  defp normalize_source_beam(options, config) do
+    compile_path =
+      if Mix.Project.umbrella?(config) do
+        umbrella_compile_paths()
+      else
+        Mix.Project.compile_path
+      end
+
+    Keyword.put_new(options, :source_beam, compile_path)
+  end
+
+  defp umbrella_compile_paths do
+    Enum.map(umbrella_apps(), fn dep ->
+      dest = Keyword.fetch!(dep.opts, :dest)
+      Mix.Project.in_project(dep.app, dest, fn module ->
+        module.project
+        |> Keyword.put_new(:build_per_environment, true)
+        |> Mix.Project.compile_path()
+      end)
+    end)
+  end
+
+  defp umbrella_apps do
+    Mix.Dep.loaded([])
```

Comment:
`Mix.Dep` is not public API so I would prefer to use `Mix.Dep.Umbrella.unloaded` which also not public API but would at least give correct results. Once you do that, you can look at `Path.join dep.opts[:build], "ebin"` for the compile path.